### ISSUE TITLE
'AdamW' object is not callable

### ISSUE THREAD

**Issue by salorsmile** [davda54/sam#29]
how to save this problem?

def build_optimizer(config, model):
    """
    Build optimizer, set weight decay of normalization to 0 by default.
    """

    skip = {}
    skip_keywords = {}
    if hasattr(model, 'no_weight_decay'):
        skip = model.no_weight_decay()
    if hasattr(model, 'no_weight_decay_keywords'):
        skip_keywords = model.no_weight_decay_keywords()
    parameters = set_weight_decay(model, skip, skip_keywords)

    opt_lower = config.TRAIN.OPTIMIZER.NAME.lower()
    optimizer = None
    if opt_lower == 'sgd':
        optimizer = optim.SGD(parameters, momentum=config.TRAIN.OPTIMIZER.MOMENTUM, nesterov=True,
                              lr=config.TRAIN.BASE_LR, weight_decay=config.TRAIN.WEIGHT_DECAY)
    elif opt_lower == 'adamw':
        optimizer = optim.AdamW(parameters, eps=config.TRAIN.OPTIMIZER.EPS, betas=config.TRAIN.OPTIMIZER.BETAS,
                                lr=config.TRAIN.BASE_LR, weight_decay=config.TRAIN.WEIGHT_DECAY)

    optimizer_sam = SAM(parameters, optimizer, momentum=0.9)

    return optimizer_sam

**Comment by davda54** [comment]
Hi,

the base optimizer is initialized inside of `SAM` and that's why it doesn't work in your case (`optimizer` needs to be callable to be initialized). You can try something like:

```python
optimizer_sam = SAM(parameters, optim.AdamW, eps=config.TRAIN.OPTIMIZER.EPS, betas=config.TRAIN.OPTIMIZER.BETAS, lr=config.TRAIN.BASE_LR, weight_decay=config.TRAIN.WEIGHT_DECAY)
```

**Comment by salorsmile** [comment]
it works, thank u.
